File: src/LooperAnalyzer/Analysis/OptimizationCandidateAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LooperAnalyzer.Analysis
{
    /// <summary>
    /// Extracts candidates for optimization.
    /// </summary>
    internal sealed class OptimizationCandidateAnalyzer : CSharpSyntaxWalker // TODO: hide walker
    {
        SemanticModel _model;
        List<OptimizationCandidate> _candidates;
        List<InvalidMarkedNode> _falselyMarkedNodes;
        List<OptimizedNode> _optimizedNodes;

        /// <summary>
        /// Expressions or statements that can be possibly optimized.
        /// </summary>
        public IEnumerable<OptimizationCandidate> Candidates => _candidates;

        /// <summary>
        /// Get SyntaxNodes that have been marked for optimization but are not valid candidates.
        /// </summary>
        public IEnumerable<InvalidMarkedNode> FalselyMarkedNodes => _falselyMarkedNodes;

        /// <summary>
        /// Get nodes that are marked for optimization and code has been generated.
        /// </summary>
        public IEnumerable<MarkedNode> OptimizedNodes => _optimizedNodes;

        public OptimizationCandidateAnalyzer(SemanticModel model)
        {
            _model = model;
            _candidates = new List<OptimizationCandidate>();
            _falselyMarkedNodes = new List<InvalidMarkedNode>();
            _optimizedNodes = new List<OptimizedNode>();
        }

        public void Run() => Visit(_model.SyntaxTree.GetRoot());

        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
        {
            // Get parent stmt to check if it's marked
            var stmt = node.GetParentStatement();
            var isMarkedForOptimization = stmt.IsMarkedWithOptimizationTrivia();

            // If method invoked is a Linq consumer method and is not marked for optimization 
            // add to candidate list, else add to marked nodes.
            // TODO: check if generated code is valid.
            var member = node.Expression as MemberAccessExpressionSyntax;
            var method = _model.GetSymbolInfo(member).Symbol as IMethodSymbol;
            if (method?.IsOptimizableConsumerMethod() == true && _model.GetTypeInfo(member.Expression).Type.IsOptimizableSourceType()) {
                if (!isMarkedForOptimization == true) {
                    _candidates.Add(OptimizationCandidate.FromInvocation(node));
                    return;
                }
                else {
                    //_markedNodes.Add();
                    return;
                }
            }

            // If method is marked but isn't valid
            if (isMarkedForOptimization) {
                var n = method.IsLinqMethod() 
                    ? InvalidMarkedNode.NoConsumer(node) 
                    : InvalidMarkedNode.InvalidLinq(node);
                _falselyMarkedNodes.Add(n);
            }

            base.VisitInvocationExpression(node);
        }
    }
}

File: src/Looper/Analysis/OptimizationCandidateAnalyzer.fs
namespace Looper.Core

open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp
open Microsoft.CodeAnalysis.CSharp.Syntax
open System.Collections.Generic

type OptimizationCandidateAnalyzer private (model: SemanticModel, candidates: List<OptimizationCandidate>, invalidNodes: List<InvalidNode>, optimizedNodes: List<OptimizedNode>) =
    inherit CSharpSyntaxWalker()

    member __.Candidates with get () = candidates :> seq<_>
    member __.FalselyMarkedNodes with get () = invalidNodes :> seq<_>
    member __.OptimizedNode with get () = optimizedNodes :> seq<_> 

    new(model: SemanticModel) = OptimizationCandidateAnalyzer(model, new List<_>(), new List<_>(), new List<_>())

    member __.Run () = __.Visit(model.SyntaxTree.GetRoot())

    override __.VisitInvocationExpression(node) =
        let stmt = node.GetParentStatement()
        let isMarked = stmt |> Option.exists(fun s -> s.IsMarkedWithOptimizationTrivia)

        match node.Expression with
        | :? MemberAccessExpressionSyntax as memberExpr ->
            let methodSym = model.GetSymbolInfo(memberExpr).Symbol :?> IMethodSymbol // TODO
            if methodSym.IsOptimizableConsumerMethod && model.GetTypeInfo(memberExpr.Expression).Type.IsOptimizableSourceType then
                if not isMarked then
                    candidates.Add(OptimizationCandidate.FromInvocation(node))
                    ()
                else
                    ()
            elif isMarked then
                let n = if methodSym.IsLinqMethod then NoConsumer(node) else InvalidExpression(node)
                invalidNodes.Add(n)
        | _ -> base.VisitInvocationExpression(node)